connaisseur/image.py

```diff
@@ -27,61 +27,63 @@ class Image:
-        self.name = name_tag.removesuffix(":" + str(self.tag)).removesuffix(
-            "@sha256:" + str(self.digest)
+    def __init__(self, image: str):  # pylint: disable=too-many-locals
+
+        # implements https://github.com/distribution/distribution/blob/main/reference/regexp.go
+        digest_hex = r"[0-9a-fA-F]{32,}"
+        digest_algorithm_component = r"[A-Za-z][A-Za-z0-9]*"
+        digest_algorithm_separator = r"[+._-]"
+        digest_algorithm = (
+            rf"{digest_algorithm_component}(?:{digest_algorithm_separator}"
+            rf"{digest_algorithm_component})*"
         )
-
-        first_comp = image.removesuffix(name_tag).split("/")[0]
-        self.registry = (
-            first_comp
-            if re.search(r"[.:]", first_comp)
-            or first_comp == "localhost"
-            or any(ele.isupper() for ele in first_comp)
-            else "docker.io"
-        )
-        self.repository = (
-            image.removesuffix(name_tag).removeprefix(self.registry)
-        ).strip("/") or ("library" if self.registry == "docker.io" else "")
-
-        if (self.repository + self.name).lower() != self.repository + self.name:
+        digest = rf"{digest_algorithm}:{digest_hex}"
+        tag = r"[\w][\w.-]{0,127}"
+        separator = r"[_.]|__|[-]*"
+        alpha_numeric = r"[a-z0-9]+"
+        path_component = rf"{alpha_numeric}(?:{separator}{alpha_numeric})*"
+        port = r"[0-9]+"
+        domain_component = r"(?:[a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9-]*[a-zA-Z0-9])"
+        domain_name = rf"{domain_component}(?:\.{domain_component})*"
+        ipv6 = r"\[(?:[a-fA-F0-9:]+)\]"
+        host = rf"(?:{domain_name}|{ipv6})"
+        domain = rf"{host}(?::{port})?"
+        name = rf"(?:{domain}/)?{path_component}(?:/{path_component})*"
+        reference = rf"^(?P<name>{name})(?::(?P<tag>{tag}))?(?:@(?P<digest>{digest}))?$"
+
+        match = re.search(reference, image)
+        if (not match) or (len(match.group("name")) > 255):
             msg = "{image} is not a valid image reference."
             raise InvalidImageFormatError(message=msg, image=image)
 
-    def set_digest(self, digest):
-        """
-        Set the digest to the given `digest`.
-        """
-        self.digest = digest
-
-    def has_digest(self) -> bool:
-        """
-        Return `True` if the image has a digest, `False` otherwise.
-        """
-        return self.digest is not None
+        name, tag, digest = match.groups()
+        self.digest = digest.split(":")[-1] if digest else None
+        self.tag = tag or ("latest" if not self.digest else None)
+        components = name.split("/")
+        self.name = components[-1]
+        registry_repo = components[:-1]
+
+        try:
+            registry = registry_repo[0]
+            self.registry = (
+                registry
+                if re.search(r"[.:]", registry)
+                or registry == "localhost"
+                or any(ele.isupper() for ele in registry)
```

Comment:
What is the reason here?

Reply:
that's remnants from a spec we previously implemented. i just kept it that way.